torch/nn/utils/prune.py

```diff
@@ -0,0 +1,1269 @@
+
+            Args:
+                method (a BasePruningMethod subclass): pruning method
+                    currently being applied.
+                t (torch.Tensor): tensor representing the parameter to prune
+                    (of same dimensions as mask).
+                mask (torch.Tensor): mask from previous pruning iteration
+
+            Returns:
+                new_mask (torch.Tensor): new mask that combines the effects
+                    of the old mask and the new mask from the current 
+                    pruning method (of same dimensions as mask and t).
+            """
+            new_mask = mask  # start off from existing mask
+            new_mask = new_mask.to(dtype=t.dtype)
+
+            # compute a slice of t onto which the new pruning method will operate
+            if method.PRUNING_TYPE == "unstructured":
+                # prune entries of t where the mask is 1
+                slc = mask == 1
+
+            # for struct pruning, exclude channels that have already been
+            # entirely pruned
+            elif method.PRUNING_TYPE == "structured":
+                if not hasattr(method, "axis"):
+                    raise AttributeError(
+                        "Pruning methods of PRUNING_TYPE "
+                        '"structured" need to have the attribute `axis` defined.'
+                    )
+
+                # find the channels to keep by removing the ones that have been
+                # zeroed out already (i.e. where sum(entries) == 0)
+                n_dims = len(t.shape)  # "is this a 2D tensor? 3D? ..."
+                axis = method.axis
+                if axis == -1:
+                    axis = n_dims - 1  # -1 means last dimension
+                # find channels along dim = axis that aren't already tots 0ed out
+                keep_channel = (
+                    mask.sum(dim=[d for d in range(n_dims) if d != axis]) != 0
+                )
+                # create slice to identify what to prune
+                slc = [slice(None)] * n_dims
+                slc[axis] = keep_channel
+
+            elif method.PRUNING_TYPE == "global":
+                n_dims = len(t.shape)  # "is this a 2D tensor? 3D? ..."
+                slc = [slice(None)] * n_dims
+
+            else:
+                raise ValueError(
+                    "Unrecognized PRUNING_TYPE {}".format(method.PRUNING_TYPE)
+                )
+
+            # compute the new mask on the unpruned slice of the tensor t
+            partial_mask = method.compute_mask(t[slc], default_mask=mask[slc])
+            new_mask[slc] = partial_mask.to(dtype=new_mask.dtype)
+
+            return new_mask
+
+        # apply the latest method by combining the default mask with
+        # the new mask that it generates
+        method = self._pruning_methods[-1]
+        mask = _combine_masks(method, t, default_mask)
+        return mask
+
+
+class IdentityPruningMethod(BasePruningMethod):
+    r"""Doesn't prune any units.
```

Comment:
No, it's useful to create the pruning reparametrization in a module (so to have the mask buffer and the `<paramname>_orig` parameter), even though the mask is all ones. It's so that one can put the network in a pruned state even without actively pruning anything yet.

Reply:
> It's so that one can put the network in a pruned state even without actively pruning anything yet.

Right, what is the use-case for it?

Reply:
If you need the reparametrization to be present, perhaps to reload the state easily from a pruned model, or just to compare two models (one pruned, one unpruned) using the same tensor names and analysis code. Basically just to standardize access to info across models and avoid code that looks like `if pruned: access this; else: access that`. 

If you don't find it useful and nice to have, it can be removed, but I think it's convenient to have (biased point of view).

Reply:
i don't mind it being there, I just wanted clarity on why it's there.